packages/core-app-api/src/routing/RouteTracker.tsx

```diff
@@ -0,0 +1,109 @@
+/*
+ * Copyright 2021 The Backstage Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import React, { useEffect, useMemo } from 'react';
+import { matchRoutes, useLocation } from 'react-router-dom';
+import {
+  useAnalytics,
+  AnalyticsContext,
+  CommonAnalyticsContext,
+  RouteRef,
+} from '@backstage/core-plugin-api';
+import { routeObjectCollector } from './collectors';
+import {
+  childDiscoverer,
+  routeElementDiscoverer,
+  traverseElementTree,
+} from '../extensions/traversal';
+import { BackstageRouteObject } from './types';
+
+/**
+ * Returns an extension context given the current pathname and a list of
+ * Backstage route objects.
+ */
+const getExtensionContext = (
+  pathname: string,
+  routes: BackstageRouteObject[],
+): CommonAnalyticsContext | {} => {
+  try {
+    const matches = matchRoutes(routes, { pathname });
+    const routeObject = matches
+      ?.filter(
+        match => (match?.route as BackstageRouteObject).routeRefs?.size > 0,
+      )
+      .pop()?.route as BackstageRouteObject;
```

Comment:
These lines are not clear to me on first reading, can you explain? Maybe add the explanation in a code comment or variable name?

Reply:
Updated with some docs (and a little fix).  Thanks!